src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+                    }
+
+                    // If no data was made available, we must be at the end of the stream
+                    Debug.Assert(_responseComplete);
+                    return 0;
+                }
+            }
+
+            public ValueTask<int> ReadResponseBodyAsync(Memory<byte> buffer)
+            {
+                Debug.Assert(buffer.Length > 0);
+
+                Task onDataAvailable;
+                lock (_responseBufferLock)
+                {
+                    if (_responseBuffer.ReadBytes.Length > 0)
+                    {
+                        return new ValueTask<int>(ReadFromBuffer(buffer.Span));
+                    }
+
+                    if (_responseComplete)
+                    {
+                        return new ValueTask<int>(0);
+                    }
+
+                    Debug.Assert(_responseDataAvailable == null);
+                    _responseDataAvailable = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    onDataAvailable = _responseDataAvailable.Task;
+                }
+
+                return ReadResponseBodyAsyncCore(onDataAvailable, buffer);
+            }
+
+            protected virtual void Dispose(bool disposing)
+            {
+                // TODO: Make this interlocked
+                if (!_disposed)
+                {
+                    if (disposing)
+                    {
+                        if (!_connection._httpStreams.TryRemove(_streamId, out Http2Stream removed))
+                        {
+                            Debug.Fail("_httpStreams.TryRemove failed");
+                        }
+
+                        Debug.Assert(removed == this, "_httpStreams.TryRemove returned unexpected stream");
+                    }
+
+                    _disposed = true;
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+            }
+        }
+
+        // TODO: Refactor this so that we can share a common base class with HTTP/1.1 content streams.
+        class Http2ResponseStream : Stream
+        {
+            private readonly Http2Stream _http2Stream;
+
+            public Http2ResponseStream(Http2Stream http2Stream)
+            {
+                _http2Stream = http2Stream;
+            }
+
+            public override bool CanRead => true;
+
+            public override bool CanSeek => false;
+
+            public override bool CanWrite => false;
+
+            public override long Length => throw new NotImplementedException();
+
+            public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
+
+            public override void Flush()
+            {
+                throw new NotImplementedException();
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                throw new NotImplementedException();
```

Comment:
Presumably this should be ReadAsync(...).GetAwaiter().GetResult()?

Reply:
see above